### PR TITLE
feat: Add new plugin `Album actions`

### LINKED ISSUE #831: [Feature Request] Like all songs in an Album
I think the title says it all. I think it would be great to have a "like all songs" option for an album, instead of the "add to library" option which is currently there. I like to shuffle my "Liked songs" every once in a while, so I think this will be better than to play the albums I like one-by-one.

### PR DESCRIPTION
**Pull request by MiepHD**
Adds Undislike, Dislike, Like and Unlike buttons to apply this to all songs in a playlist or album.
image
Fixes #831

### CODE REVIEW

**Comment by ArjixWasTaken** [comment]
Don't add `enabled: true` in the config....
Your plugin must not be enabled unless specified otherwise by the user's config.

**Comment by MiepHD** [comment]
I enabled it by default because it just adds these buttons to the UI. If you don't click the buttons nothing happens. But I can remove that if that's a problem

**Comment by ArjixWasTaken** [comment]
Since it does not fix a bug, I consider it an additional feature.
I guess I'll leave it up to @JellyBrick

**Comment by JellyBrick** [comment]
image

The shapes of the buttons can be confusing to users.

**Comment by MiepHD** [comment]
Why? Should I add an explanation what they do?

**Comment by JellyBrick** [comment]
What about showing just one of each button? (1 dislike button, 1 like button)
- If any item is "Liked", show a filled "Like" button; if any item is "Disliked", show a filled "Dislike" button
- If both "Liked" and "Disliked" are checked, show both filled buttons

**Comment by MiepHD** [comment]
Now it shows a 
- undislike button if any item is disliked
- dislike button if any item isn't disliked
- like button if any item isn't liked
- unlike button if any item is liked

File: src/plugins/album-actions/index.ts [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+  name: () => t('plugins.album-actions.name'),
+  description: () => t('plugins.album-actions.description'),
+  restartNeeded: false,
+  config: {
+    enabled: true,
+  },
+  renderer: {
+    observer: null as MutationObserver | null,
+    loadobserver: null as MutationObserver | null,
+    changeobserver: null as MutationObserver | null,
```
Comment (JellyBrick) [inline src/plugins/album-actions/index.ts]: use camelCase

File: src/plugins/album-actions/index.ts [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+      this.waitForElem('#continuations').then(() => {
+        const buttons: Array<HTMLElement> = [
+          ElementFromHtml(undislikeHTML),
+          ElementFromHtml(dislikeHTML),
+          ElementFromHtml(likeHTML),
+          ElementFromHtml(unlikeHTML),
+        ];
+        const playlist = document.querySelector('ytmusic-shelf-renderer')
+          ? document.querySelector('ytmusic-shelf-renderer')
+          : document.querySelector('ytmusic-playlist-shelf-renderer');
```
Comment (JellyBrick) [inline src/plugins/album-actions/index.ts]:
```suggestion
        const playlist = document.querySelector('ytmusic-shelf-renderer') ?? document.querySelector('ytmusic-playlist-shelf-renderer');
```

File: src/plugins/album-actions/index.ts [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+      loader?.style.setProperty('left', '50%');
+      loader?.style.setProperty('position', 'absolute');
+    },
+    applyToList(id: string, loader: HTMLElement) {
+      if (loader.children.length != 0) return;
+      this.loadobserver?.disconnect();
+      let playlistbuttons: NodeListOf<Element> | undefined;
+      const playlist = document.querySelector('ytmusic-shelf-renderer')
+        ? document.querySelector('ytmusic-shelf-renderer')
+        : document.querySelector('ytmusic-playlist-shelf-renderer');
```
Comment (JellyBrick) [inline src/plugins/album-actions/index.ts:133]: ditto

File: src/plugins/album-actions/index.ts [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+            '#button-shape-like[aria-pressed=true] > button',
+          );
+          break;
+        default:
+      }
+      if (playlistbuttons) {
+        for (const elem of playlistbuttons) {
+          elem.click();
+        }
+      }
```
Comment (JellyBrick) [inline src/plugins/album-actions/index.ts]:
```suggestion
      playlistButtons?.forEach((elem) => elem.click());
```

File: src/plugins/album-actions/index.ts [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+            playlist?.querySelectorAll(
+              '#button-shape-like[aria-pressed=false] > button',
+            ).length,
+            playlist?.querySelectorAll(
+              '#button-shape-like[aria-pressed=true] > button',
+            ).length,
+          ];
+          let i = 0;
+          for (const count of counts) {
+            console.log(count);
```
Comment (JellyBrick) [inline src/plugins/album-actions/index.ts]:
```suggestion
```

File: src/plugins/album-actions/index.ts [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+            if (count == 0) {
+              buttons.splice(i, 1);
+              i--;
+            }
+            i++;
+          }
+        }
```
Comment (JellyBrick) [inline src/plugins/album-actions/index.ts]: How about `.filter()` instead of this logic?

**Review by JellyBrick — CHANGES_REQUESTED** [review]

**Comment by inson1** [comment]
> Now it shows a
> 
> * undislike button if any item is disliked
> * dislike button if any item isn't disliked
> * like button if any item isn't liked
> * unlike button if any item is liked

Just another opinion. I hope you dont mind it :) That is standard, yes. But it isnt perfect :/. 

What about highlighted (bold lines?) like/dislike button if some are liked/disliked, but not all? (I think that is also used somewhere) Or something like that? Maybe like/dislike button with full white, but black "ball" inside
Just something so normal user would get it and "profi" user would know if all songs are liked or disliked.
Or another mode for this?

File: src/plugins/album-actions/index.ts [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+    waitForElem(selector: string) {
+      return new Promise((resolve) => {
+        const interval = setInterval(() => {
+          const elem = document.querySelector(selector);
+          if (!elem) return;
+
+          clearInterval(interval);
+          resolve(elem);
+        });
+      });
```
Comment (Su-Yong) [inline src/plugins/album-actions/index.ts:175]:
```suggestion
    waitForElem(selector: string) {
      let limit = 0;
      return new Promise((resolve, reject) => {
        const interval = setInterval(() => {
          limit += 1;
          if (limit > 3000) {
            clearInterval(interval);
            reject();
          }
          const elem = document.querySelector(selector);
          if (!elem) return;

          clearInterval(interval);
          resolve(elem);
        });
      });
```

File: src/plugins/album-actions/index.ts [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+          }
+        }
+        const menu = document.querySelector('.detail-page-menu');
+        if (menu && !document.querySelector('.like-menu')) {
+          for (const button of buttons) {
+            menu.appendChild(button);
+            button.addEventListener('click', this.loadFullList);
+          }
+        }
+      });
```
Comment (Su-Yong) [inline src/plugins/album-actions/index.ts:108]:
```suggestion
      }).catch(() => {
        // Cannot find element
      });
```

**Review by Su-Yong — COMMENTED** [review]
It seems memory leak bug

**Review by Su-Yong — APPROVED** [review]

**Review by JellyBrick — APPROVED** [review]

**Pull request merged**